File: cpp/test/test_wire.cpp
#include <doctest/doctest.h>
#include <nlohmann/json.hpp>

#include <embedded_bridge/detail/crc16.h>
#include <embedded_bridge/framing/hdlc.h>
#include <embedded_bridge/framing/slip.h>
#include <embedded_bridge/framing/cobs.h>
#include <embedded_bridge/message.h>

#include <fstream>
#include <string>
#include <vector>

using json = nlohmann::json;
using namespace ebridge;

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

static std::vector<uint8_t> hex_to_bytes(const std::string& hex) {
    std::vector<uint8_t> out;
    for (size_t i = 0; i + 1 < hex.size(); i += 2) {
        out.push_back(static_cast<uint8_t>(
            std::stoul(hex.substr(i, 2), nullptr, 16)));
    }
    return out;
}

static std::string bytes_to_hex(const uint8_t* data, size_t len) {
    static const char digits[] = "0123456789abcdef";
    std::string out;
    out.reserve(len * 2);
    for (size_t i = 0; i < len; i++) {
        out.push_back(digits[data[i] >> 4]);
        out.push_back(digits[data[i] & 0x0F]);
    }
    return out;
}

static json load_vectors(const char* filename) {
    std::string path = std::string(WIRE_TESTS_DIR) + "/" + filename;
    std::ifstream f(path);
    REQUIRE(f.good());
    json j;
    f >> j;
    return j;
}

/// Helper to get raw bytes from a BufferWriter (which stores as char).
static std::string writer_to_hex(const BufferWriter<256>& w) {
    return bytes_to_hex(reinterpret_cast<const uint8_t*>(w.str()), w.len());
}

/// Feed a byte vector into a framer one byte at a time.
template <typename Framer>
static void feed_bytes(Framer& framer, const std::vector<uint8_t>& data) {
    for (uint8_t b : data) {
        framer.process_byte(b);
    }
}

/// Callback context for framers.
struct FrameCapture {
    std::vector<uint8_t> last_frame;
    int frame_count = 0;

    static void callback(void* ctx, const uint8_t* data, size_t len) {
        auto* self = static_cast<FrameCapture*>(ctx);
        self->last_frame.assign(data, data + len);
        self->frame_count++;
    }
};

// ---------------------------------------------------------------------------
// CRC-16
// ---------------------------------------------------------------------------

TEST_CASE("wire: crc16") {
    auto j = load_vectors("crc16.json");

    SUBCASE("vectors") {
        for (const auto& v : j["vectors"]) {
            CAPTURE(v["name"].get<std::string>());
            auto input = hex_to_bytes(v["input_hex"].get<std::string>());
            uint16_t expected = v["crc"].get<uint16_t>();
            uint16_t actual = detail::crc16_hdlc(input.data(), input.size());
            CHECK(actual == expected);
        }
    }

    SUBCASE("residue") {
        for (const auto& v : j["residue_vectors"]) {
            CAPTURE(v["name"].get<std::string>());
            auto frame = hex_to_bytes(v["frame_hex"].get<std::string>());
            uint16_t expected = v["frame_crc"].get<uint16_t>();
            uint16_t actual = detail::crc16_hdlc(frame.data(), frame.size());
            CHECK(actual == expected);
            CHECK(actual == j["good_residue"].get<uint16_t>());
        }
    }
}

// ---------------------------------------------------------------------------
// HDLC
// ---------------------------------------------------------------------------

TEST_CASE("wire: hdlc encode/decode") {
    auto j = load_vectors("hdlc.json");

    for (const auto& v : j["encode_vectors"]) {
        CAPTURE(v["name"].get<std::string>());
        auto payload = hex_to_bytes(v["payload_hex"].get<std::string>());

        // Encode
        BufferWriter<256> buf;
        HdlcFramingWriter<256> writer(buf);
        writer.write(payload.data(), payload.size());
        writer.end_frame();
        CHECK(writer_to_hex(buf) == v["encoded_hex"].get<std::string>());

        // Decode (roundtrip)
        auto wire = hex_to_bytes(v["encoded_hex"].get<std::string>());
        FrameCapture cap;
        HdlcFramer<256> framer(FrameCapture::callback, &cap);
        feed_bytes(framer, wire);
        CHECK(cap.frame_count == 1);
        CHECK(cap.last_frame == payload);
    }
}

TEST_CASE("wire: hdlc decode errors") {
    auto j = load_vectors("hdlc.json");

    for (const auto& v : j["decode_error_vectors"]) {
        CAPTURE(v["name"].get<std::string>());
        auto wire = hex_to_bytes(v["wire_hex"].get<std::string>());

        FrameCapture cap;
        HdlcFramer<256> framer(FrameCapture::callback, &cap);
        feed_bytes(framer, wire);
        CHECK(cap.frame_count == 0);
    }
}

// ---------------------------------------------------------------------------
// SLIP
// ---------------------------------------------------------------------------

TEST_CASE("wire: slip encode/decode") {
    auto j = load_vectors("slip.json");

    for (const auto& v : j["encode_vectors"]) {
        CAPTURE(v["name"].get<std::string>());
        auto payload = hex_to_bytes(v["payload_hex"].get<std::string>());

        // Encode
        BufferWriter<256> buf;
        SlipFramingWriter<256> writer(buf);
        writer.write(payload.data(), payload.size());
        writer.end_frame();
        CHECK(writer_to_hex(buf) == v["encoded_hex"].get<std::string>());

        // Decode (roundtrip)
        auto wire = hex_to_bytes(v["encoded_hex"].get<std::string>());
        FrameCapture cap;
        SlipFramer<256> framer(FrameCapture::callback, &cap);
        feed_bytes(framer, wire);
        CHECK(cap.frame_count == 1);
        CHECK(cap.last_frame == payload);
    }
}

// ---------------------------------------------------------------------------
// COBS
// ---------------------------------------------------------------------------

TEST_CASE("wire: cobs encode/decode") {
    auto j = load_vectors("cobs.json");

    for (const auto& v : j["encode_vectors"]) {
        CAPTURE(v["name"].get<std::string>());
        auto payload = hex_to_bytes(v["payload_hex"].get<std::string>());

        // Encode (framed = cobs + delimiter)
        BufferWriter<256> buf;
        CobsFramingWriter<256> writer(buf);
        writer.write(payload.data(), payload.size());
        writer.end_frame();
        CHECK(writer_to_hex(buf) == v["framed_hex"].get<std::string>());

        // Decode (roundtrip)
        auto wire = hex_to_bytes(v["framed_hex"].get<std::string>());
        FrameCapture cap;
        CobsFramer<256> framer(FrameCapture::callback, &cap);
        feed_bytes(framer, wire);
        CHECK(cap.frame_count == 1);
        CHECK(cap.last_frame == payload);
    }
}

// ---------------------------------------------------------------------------
// Message protocol — varint
// ---------------------------------------------------------------------------

TEST_CASE("wire: varint encode/decode") {
    auto j = load_vectors("message.json");

    for (const auto& v : j["varint_vectors"]) {
        uint32_t value = v["value"].get<uint32_t>();
        auto expected = hex_to_bytes(v["encoded_hex"].get<std::string>());
        size_t expected_bytes = v["bytes"].get<size_t>();
        CAPTURE(value);

        // Encode
        uint8_t buf[5];
        size_t n = encode_varint(value, buf);
        CHECK(n == expected_bytes);
        CHECK(bytes_to_hex(buf, n) == v["encoded_hex"].get<std::string>());

        // Decode
        uint32_t decoded_value = 0;
        size_t consumed = 0;
        bool ok = decode_varint(expected.data(), expected.size(),
                                decoded_value, consumed);
        CHECK(ok);
        CHECK(decoded_value == value);
        CHECK(consumed == expected_bytes);
    }
}

// ---------------------------------------------------------------------------
// Message protocol — writer output
// ---------------------------------------------------------------------------

TEST_CASE("wire: message writer text") {
    auto j = load_vectors("message.json");

    for (const auto& v : j["text_vectors"]) {
        CAPTURE(v["name"].get<std::string>());
        std::string text = v["text"].get<std::string>();

        BufferWriter<256> buf;
        MessageWriter writer(buf);
        writer.write_text(text.c_str(), text.size());
        CHECK(writer_to_hex(buf) == v["wire_hex"].get<std::string>());
    }
}

TEST_CASE("wire: message writer binary") {
    auto j = load_vectors("message.json");

    for (const auto& v : j["binary_vectors"]) {
        CAPTURE(v["name"].get<std::string>());
        auto payload = hex_to_bytes(v["payload_hex"].get<std::string>());

        BufferWriter<256> buf;
        MessageWriter writer(buf);
        writer.write_binary(payload.data(), payload.size());
        CHECK(writer_to_hex(buf) == v["wire_hex"].get<std::string>());
    }
}

// ---------------------------------------------------------------------------
// Message protocol — reader
// ---------------------------------------------------------------------------

TEST_CASE("wire: message reader") {
    auto j = load_vectors("message.json");

    struct Result {
        std::string type;
        std::string value;
    };

    for (const auto& v : j["reader_vectors"]) {
        CAPTURE(v["name"].get<std::string>());
        auto wire = hex_to_bytes(v["wire_hex"].get<std::string>());

        std::vector<Result> results;

        class Capture : public BufferingMessageHandler<512, 4096> {
        public:
            std::vector<Result>& results;
            explicit Capture(std::vector<Result>& r) : results(r) {}
            void on_text(const char* data, size_t len) override {
                if (len > 0 && data[len - 1] == '\r') len--;
                results.push_back({"text", std::string(data, len)});
            }
            void on_binary(const uint8_t* data, size_t len) override {
                results.push_back({"binary", bytes_to_hex(data, len)});
            }
        };

        Capture handler(results);
        MessageReader reader(handler);
        reader.feed(wire.data(), wire.size());

        const auto& expected = v["expected"];
        REQUIRE(results.size() == expected.size());
        for (size_t i = 0; i < results.size(); i++) {
            CHECK(results[i].type == expected[i]["type"].get<std::string>());
            if (results[i].type == "text") {
                CHECK(results[i].value == expected[i]["value"].get<std::string>());
            } else {
                CHECK(results[i].value == expected[i]["value_hex"].get<std::string>());
            }
        }
    }
}

File: python/tests/test_wire.py
"""Cross-language wire compatibility tests.

Loads shared test vectors from wire-tests/ and validates the Python
implementations produce identical results to the C++ implementations.
"""

from __future__ import annotations

import json
from pathlib import Path

import pytest

from embedded_bridge.framing.crc16 import crc16_hdlc, CRC16_GOOD
from embedded_bridge.framing.hdlc import HdlcFramer, HdlcFrameEncoder
from embedded_bridge.framing.slip import SlipFramer, SlipFrameEncoder
from embedded_bridge.framing.cobs import CobsFramer, CobsFrameEncoder, cobs_encode
from embedded_bridge.framing.message import (
    MessageReader,
    MessageWriter,
    encode_varint,
    decode_varint,
    SOH,
    BINARY_PROTOCOL_V1,
)

WIRE_TESTS_DIR = Path(__file__).parents[2] / "wire-tests"


def load_vectors(filename: str) -> dict:
    with open(WIRE_TESTS_DIR / filename) as f:
        return json.load(f)


# ---------------------------------------------------------------------------
# CRC-16
# ---------------------------------------------------------------------------


class TestCRC16Wire:
    vectors = load_vectors("crc16.json")

    @pytest.mark.parametrize(
        "v", vectors["vectors"], ids=lambda v: v["name"]
    )
    def test_crc(self, v: dict) -> None:
        data = bytes.fromhex(v["input_hex"])
        assert crc16_hdlc(data) == v["crc"]

    @pytest.mark.parametrize(
        "v", vectors["residue_vectors"], ids=lambda v: v["name"]
    )
    def test_residue(self, v: dict) -> None:
        frame = bytes.fromhex(v["frame_hex"])
        assert crc16_hdlc(frame) == v["frame_crc"]
        assert crc16_hdlc(frame) == self.vectors["good_residue"]


# ---------------------------------------------------------------------------
# HDLC
# ---------------------------------------------------------------------------


class TestHDLCWire:
    vectors = load_vectors("hdlc.json")

    @pytest.mark.parametrize(
        "v", vectors["encode_vectors"], ids=lambda v: v["name"]
    )
    def test_encode(self, v: dict) -> None:
        payload = bytes.fromhex(v["payload_hex"])
        encoded = HdlcFrameEncoder.encode(payload)
        assert encoded.hex() == v["encoded_hex"]

    @pytest.mark.parametrize(
        "v", vectors["encode_vectors"], ids=lambda v: v["name"]
    )
    def test_decode_roundtrip(self, v: dict) -> None:
        payload = bytes.fromhex(v["payload_hex"])
        wire = bytes.fromhex(v["encoded_hex"])
        frames: list[bytes] = []
        framer = HdlcFramer(lambda data: frames.append(data))
        framer.process_bytes(wire)
        assert len(frames) == 1
        assert frames[0] == payload

    @pytest.mark.parametrize(
        "v", vectors["decode_error_vectors"], ids=lambda v: v["name"]
    )
    def test_decode_error(self, v: dict) -> None:
        wire = bytes.fromhex(v["wire_hex"])
        frames: list[bytes] = []
        framer = HdlcFramer(lambda data: frames.append(data))
        framer.process_bytes(wire)
        assert len(frames) == 0


# ---------------------------------------------------------------------------
# SLIP
# ---------------------------------------------------------------------------


class TestSLIPWire:
    vectors = load_vectors("slip.json")

    @pytest.mark.parametrize(
        "v", vectors["encode_vectors"], ids=lambda v: v["name"]
    )
    def test_encode(self, v: dict) -> None:
        payload = bytes.fromhex(v["payload_hex"])
        encoded = SlipFrameEncoder.encode(payload)
        assert encoded.hex() == v["encoded_hex"]

    @pytest.mark.parametrize(
        "v", vectors["encode_vectors"], ids=lambda v: v["name"]
    )
    def test_decode_roundtrip(self, v: dict) -> None:
        payload = bytes.fromhex(v["payload_hex"])
        wire = bytes.fromhex(v["encoded_hex"])
        frames: list[bytes] = []
        framer = SlipFramer(lambda data: frames.append(data))
        framer.process_bytes(wire)
        assert len(frames) == 1
        assert frames[0] == payload


# ---------------------------------------------------------------------------
# COBS
# ---------------------------------------------------------------------------


class TestCOBSWire:
    vectors = load_vectors("cobs.json")

    @pytest.mark.parametrize(
        "v", vectors["encode_vectors"], ids=lambda v: v["name"]
    )
    def test_encode_raw(self, v: dict) -> None:
        payload = bytes.fromhex(v["payload_hex"])
        encoded = cobs_encode(payload)
        assert encoded.hex() == v["cobs_hex"]

    @pytest.mark.parametrize(
        "v", vectors["encode_vectors"], ids=lambda v: v["name"]
    )
    def test_encode_framed(self, v: dict) -> None:
        payload = bytes.fromhex(v["payload_hex"])
        framed = CobsFrameEncoder.encode(payload)
        assert framed.hex() == v["framed_hex"]

    @pytest.mark.parametrize(
        "v", vectors["encode_vectors"], ids=lambda v: v["name"]
    )
    def test_decode_roundtrip(self, v: dict) -> None:
        payload = bytes.fromhex(v["payload_hex"])
        wire = bytes.fromhex(v["framed_hex"])
        frames: list[bytes] = []
        framer = CobsFramer(lambda data: frames.append(data))
        framer.process_bytes(wire)
        assert len(frames) == 1
        assert frames[0] == payload


# ---------------------------------------------------------------------------
# Message protocol — varint
# ---------------------------------------------------------------------------


class TestVarintWire:
    vectors = load_vectors("message.json")

    @pytest.mark.parametrize(
        "v", vectors["varint_vectors"], ids=lambda v: str(v["value"])
    )
    def test_encode(self, v: dict) -> None:
        encoded = encode_varint(v["value"])
        assert encoded.hex() == v["encoded_hex"]
        assert len(encoded) == v["bytes"]

    @pytest.mark.parametrize(
        "v", vectors["varint_vectors"], ids=lambda v: str(v["value"])
    )
    def test_decode(self, v: dict) -> None:
        data = bytes.fromhex(v["encoded_hex"])
        value, consumed = decode_varint(data)
        assert value == v["value"]
        assert consumed == v["bytes"]


# ---------------------------------------------------------------------------
# Message protocol — writer
# ---------------------------------------------------------------------------


class TestMessageWriterWire:
    vectors = load_vectors("message.json")

    @pytest.mark.parametrize(
        "v", vectors["text_vectors"], ids=lambda v: v["name"]
    )
    def test_write_text(self, v: dict) -> None:
        output = bytearray()
        writer = MessageWriter(output.extend)
        writer.write_text(v["text"])
        assert output.hex() == v["wire_hex"]

    @pytest.mark.parametrize(
        "v", vectors["binary_vectors"], ids=lambda v: v["name"]
    )
    def test_write_binary(self, v: dict) -> None:
        output = bytearray()
        writer = MessageWriter(output.extend)
        payload = bytes.fromhex(v["payload_hex"])
        writer.write_binary(payload)
        assert output.hex() == v["wire_hex"]


# ---------------------------------------------------------------------------
# Message protocol — reader
# ---------------------------------------------------------------------------


class TestMessageReaderWire:
    vectors = load_vectors("message.json")

    @pytest.mark.parametrize(
        "v", vectors["reader_vectors"], ids=lambda v: v["name"]
    )
    def test_read(self, v: dict) -> None:
        wire = bytes.fromhex(v["wire_hex"])
        reader = MessageReader()
        reader.feed(wire)
        messages = reader.drain()

        expected = v["expected"]
        assert len(messages) == len(expected)
        for msg, exp in zip(messages, expected):
            if exp["type"] == "text":
                assert isinstance(msg, str)
                assert msg == exp["value"]
            else:
                assert isinstance(msg, bytes)
                assert msg.hex() == exp["value_hex"]
